handwritten/hexagon_disas_c/functions.c

```diff
@@ -6,3 +6,255 @@ static inline bool is_last_instr(const ut8 parse_bits) {
+		if (!m->bits) {
+			break;
+		}
+		r |= ((val >> m->shift) & rz_num_bitmask(m->bits)) << off;
+		off += m->bits;
+	}
+	if (bits_total) {
+		*bits_total = off;
+	}
+	return r;
+}
+
+/**
+ * \return the index of the immediate operand used as the jump target or -1 if there is none.
+ */
+static int get_jmp_target_imm_op_index(const HexInsnTemplate *tpl) {
+	if (!(tpl->flags & HEX_INSN_TEMPLATE_FLAG_HAS_JMP_TGT)) {
+		return -1;
+	}
+	bool has_imm = false;
+	size_t i;
+	for (i = 0; i < HEX_MAX_OPERANDS; i++) {
+		const HexOpTemplate *op = &tpl->ops[i];
+		HexOpTemplateType type = op->info & HEX_OP_TEMPLATE_TYPE_MASK;
+		if (type == HEX_OP_TEMPLATE_TYPE_NONE) {
+			break;
+		}
+		if (type == HEX_OP_TEMPLATE_TYPE_IMM) {
+			has_imm = true;
+			if (op->info & HEX_OP_TEMPLATE_FLAG_IMM_PC_RELATIVE) {
+				return i;
+			}
+		}
+	}
+	return has_imm && i == 1 ? 0 : -1;
+}
+
+static void hex_disasm_with_templates(const HexInsnTemplate *tpl, HexState *state, ut32 hi_u32, RZ_INOUT HexInsn *hi, ut64 addr, HexPkt *pkt) {
+	bool print_reg_alias = rz_config_get_b(state->cfg, "plugins.hexagon.reg.alias");
+	bool show_hash = rz_config_get_b(state->cfg, "plugins.hexagon.imm.hash");
+	bool sign_nums = rz_config_get_b(state->cfg, "plugins.hexagon.imm.sign");
+	char signed_imm[HEX_MAX_OPERANDS][32];
+	// Find the right template
+	for (; tpl->id; tpl++) {
+		if ((hi_u32 & tpl->encoding.mask) == tpl->encoding.op) {
+			break;
+		}
+	}
+	if (!tpl->id) {
+		// unknown/invalid
+		return;
+	}
+	hi->instruction = tpl->id;
+	hi->opcode = hi_u32;
+	hi->parse_bits = (hi_u32 & HEX_PARSE_BITS_MASK) >> 14;
+	hi->pred = tpl->pred;
+
+	// textual disasm is built by copying tpl->syntax while inserting the ops at the right positions
+	RzStrBuf sb;
+	rz_strbuf_init(&sb);
+	size_t syntax_cur = 0;
+	size_t syntax_len = strlen(tpl->syntax);
+
+	hi->op_count = 0;
+	for (size_t i = 0; i < HEX_MAX_OPERANDS; i++) {
+		const HexOpTemplate *op = &tpl->ops[i];
+		HexOpTemplateType type = op->info & HEX_OP_TEMPLATE_TYPE_MASK;
+		if (type == HEX_OP_TEMPLATE_TYPE_NONE) {
+			break;
+		}
+
+		if (op->syntax > syntax_cur && op->syntax <= syntax_len) {
+			rz_strbuf_append_n(&sb, tpl->syntax + syntax_cur, op->syntax - syntax_cur);
+			syntax_cur = op->syntax;
+		}
+
+		hi->op_count++;
+		hi->ops[i].attr = 0;
+		switch (type) {
+		case HEX_OP_TEMPLATE_TYPE_IMM: {
+			hi->ops[i].type = HEX_OP_TYPE_IMM;
+			ut32 bits_total;
+			hi->ops[i].op.imm = hex_op_masks_extract(op->masks, hi_u32, &bits_total) << op->imm_scale;
+			hi->ops[i].shift = op->imm_scale;
+			if (op->imm_scale) {
+				hi->ops[i].attr |= HEX_OP_IMM_SCALED;
+			}
+			if (op->info & HEX_OP_TEMPLATE_FLAG_IMM_SIGNED) {
+				ut32 shift = bits_total + op->imm_scale - 1;
+				rz_warn_if_fail(shift < 64);
+				if (hi->ops[i].op.imm & (1ull << shift)) {
+					hi->ops[i].op.imm |= UT64_MAX << shift;
+				}
+			}
+			if (op->info & HEX_OP_TEMPLATE_FLAG_IMM_EXTENDABLE) {
+				hex_extend_op(state, &hi->ops[i], false, addr);
+			}
+			// text
```

Comment:
What is this saying?

Reply:
textual disassembly, will make it more verbose.